### PR TITLE
[a10-octavia v1.3][STACK-2330] device-name flaovr support

### PR DESCRIPTION
**Pull request by ytsai-a10**
## Description
document for the feature: https://teams.microsoft.com/l/file/45BFD3D5-6406-4A7F-B2C6-EACF24CA9E31?tenantId=91d27ab9-8c5e-41d4-82e8-3d1bf81fcb2f&fileType=docx&objectUrl=https%3A%2F%2Fa10networks.sharepoint.com%2Fsites%2FOpenstack%2FShared%20Documents%2FDevelopment%2FResearch%20%26%20Design%2Fa10-octavia%20v1.3%20-%20Device%20Flavor%2FDevice%20Flavor.docx&baseUrl=https%3A%2F%2Fa10networks.sharepoint.com%2Fsites%2FOpenstack&serviceName=teams&threadId=19:0f1acbb173d74758b05ee8bacb000d68@thread.tacv2&groupId=37bbf3ad-c05a-4e67-b0cc-12bcd1479beb

## Jira Ticket
https://a10networks.atlassian.net/browse/STACK-2330

## Technical Approach
document for the feature: https://teams.microsoft.com/l/file/45BFD3D5-6406-4A7F-B2C6-EACF24CA9E31?tenantId=91d27ab9-8c5e-41d4-82e8-3d1bf81fcb2f&fileType=docx&objectUrl=https%3A%2F%2Fa10networks.sharepoint.com%2Fsites%2FOpenstack%2FShared%20Documents%2FDevelopment%2FResearch%20%26%20Design%2Fa10-octavia%20v1.3%20-%20Device%20Flavor%2FDevice%20Flavor.docx&baseUrl=https%3A%2F%2Fa10networks.sharepoint.com%2Fsites%2FOpenstack&serviceName=teams&threadId=19:0f1acbb173d74758b05ee8bacb000d68@thread.tacv2&groupId=37bbf3ad-c05a-4e67-b0cc-12bcd1479beb

## Config Changes

<pre>
<b>[hardware_thunder]
devices = [
                    {
                     "project_id": "3d21c71c4e4040599933bda62a76ae2f",
                     "ip_address": "192.168.90.46",
                     "username": "admin",
                     "password": "a10",
                     "device_name": "admin"
                     },
                    {
                     "project_id": "9955a18b75fa4c30bb865d72b5a86a6a",
                     "ip_address": "192.168.90.46",
                     "username": "admin",
                     "password": "a10",
                     "partition_name": "p7",
                     "device_name": "child1",
                     "hierarchical_multitenancy": "enable"
                     },
                    {
                     "project_id": "b612261c15374f1a896f4c3830042f48",
                     "ip_address": "192.168.90.46",
                     "username": "admin",
                     "password": "a10",
                     "partition_name": "p7",
                     "device_name": "child2",
                     "hierarchical_multitenancy": "enable",
                     },
                    {
                     "ip_address": "192.168.90.46",
                     "username": "admin",
                     "password": "a10",
                     "partition_name": "p7",
                     "device_name": "dev1"
                     },
                    {
                     "ip_address": "192.168.90.45",
                     "username": "admin",
                     "password": "a10",
                     "device_name": "dev2"
                     },
                    {
                     "ip_address": "192.168.90.46",
                     "username": "admin",
                     "password": "a10",
                     "device_name": "dev3",
                     "hierarchical_multitenancy": "enable"
                     },
             ]
</b>
</pre>

## Test Cases
- flavor schema 
```shell
openstack loadbalancer provider capability list a10
```

---------------------

 - use device-name flavor
Flavor:
```json
{ 
	"device-name": "dev2"
}
```

```shell
openstack loadbalancer flavorprofile create --name fp_dev2 --provider a10 --flavor-data '{"device-name": "dev2"}'
openstack loadbalancer flavor create --name f_dev2 --flavorprofile fp_dev2 --description "use device dev2" --enable
openstack loadbalancer create --flavor f_dev2 --vip-subnet-id tp91 --vip-address 192.168.91.56 --name vip1
openstack loadbalancer listener create --protocol HTTP --protocol-port 80 --name vport1 vip1
```

---------------------

 - partition support
```
{ 
	"device-name": "dev1"
}
```

```shell
openstack loadbalancer flavorprofile create --name fp_dev1 --provider a10 --flavor-data '{"device-name": "dev1"}'
openstack loadbalancer flavor create --name f_dev1 --flavorprofile fp_dev1 --description "use device dev1" --enable

openstack loadbalancer create --flavor f_dev1 --vip-subnet-id tp91 --vip-address 192.168.91.56 --name vip1
openstack loadbalancer listener create --protocol HTTP --protocol-port 80 --name vport1 vip1
```

---------------------

 - HMT
```
{ 
	"device-name": "dev3"
}
```

```
openstack loadbalancer flavorprofile create --name fp_dev3 --provider a10 --flavor-data '{"device-name": "dev3"}'
openstack loadbalancer flavor create --name f_dev3 --flavorprofile fp_dev3 --description "use device dev1" --enable

openstack loadbalancer create --flavor f_dev3 --vip-subnet-id tp91 --vip-address 192.168.91.56 --project child1 --name vip1
openstack loadbalancer listener create --protocol HTTP --protocol-port 80 --name vport1 vip1
```

---------------------

- use parent partition

```
[a10_global]
use_parent_partition=True
```

```
openstack loadbalancer create --flavor f_dev3 --vip-subnet-id tp91 --vip-address 192.168.91.56 --project child1 --name vip1
openstack loadbalancer listener create --protocol HTTP --protocol-port 80 --name vport1 vip1
```

---------------------

## Manual Testing
- flavor schema 
```shell
stack@ytsai-openstack:~/source/a10-octavia/device-flavor$ openstack loadbalancer provider capability list a10
+---------------------------------+----------------------------------------------------------------------------------------------------+
| name                            | description                                                                                        |
+---------------------------------+----------------------------------------------------------------------------------------------------+
| server.name-expressions         | Specify name expression to match members and axapi that will apply to the server                   |
| nat-pool-list                   | Specify axapi of nat pools for loadbalancer                                                        |
| health-monitor.name-expressions | Specify name expression to match healthmonitor and axapi that will apply to the health monitor     |
| device-name                     | Specify vthunder device name that used for this loadbalancer                                       |
| virtual-server.name-expressions | Specify name expression to match loadbalancers and axapi that will apply to the slb virtual-server |
| nat-pool                        | Specify axapi of default nat pool for loadbalancer                                                 |
| virtual-port.name-expressions   | Specify name expression to match listeners and axapi that will apply to the vport                  |
| service-group.name-expressions  | Specify name expression to match pools and axapi that will apply to the service-group              |
| server                          | Specify axapi that will apply to the server                                                        |
| service-group                   | Specify axapi that will apply to the service-group                                                 |
| virtual-port                    | Specify axapi that will apply to the vport                                                         |
| virtual-server                  | Specify axapi that will apply to the slb virtual-server                                            |
| health-monitor                  | Specify axapi that will apply to the health monitor                                                |
+---------------------------------+----------------------------------------------------------------------------------------------------+
```
---------------

 - use device-name flavor
```
stack@ytsai-openstack:~/source/a10-octavia/device-flavor$ openstack loadbalancer create --flavor f_dev2 --vip-subnet-id tp91 --vip-address 192.168.91.56 --name vip1
+---------------------+--------------------------------------+
| Field               | Value                                |
+---------------------+--------------------------------------+
| admin_state_up      | True                                 |
| created_at          | 2021-05-12T10:06:54                  |
| description         |                                      |
| flavor_id           | 05084690-b8d8-4864-b1ab-017e8add7c07 |
| id                  | 1d76140c-de29-462a-bbb8-3fa98a5a3425 |
| listeners           |                                      |
| name                | vip1                                 |
| operating_status    | OFFLINE                              |
| pools               |                                      |
| project_id          | 3d21c71c4e4040599933bda62a76ae2f     |
| provider            | a10                                  |
| provisioning_status | PENDING_CREATE                       |
| updated_at          | None                                 |
| vip_address         | 192.168.91.56                        |
| vip_network_id      | 5e1b43e5-9f65-40af-a28f-88e63dc6d667 |
| vip_port_id         | 38af18b7-4eba-46fb-861b-07808e959beb |
| vip_qos_policy_id   | None                                 |
| vip_subnet_id       | f25ce642-f953-4058-8c46-98fdd72fb129 |
+---------------------+--------------------------------------+
stack@ytsai-openstack:~/source/a10-octavia/device-flavor$ openstack loadbalancer listener create --protocol HTTP --protocol-port 80 --name vport1 vip1
+-----------------------------+--------------------------------------+
| Field                       | Value                                |
+-----------------------------+--------------------------------------+
| admin_state_up              | True                                 |
| connection_limit            | -1                                   |
| created_at                  | 2021-05-12T10:07:26                  |
| default_pool_id             | None                                 |
| default_tls_container_ref   | None                                 |
| description                 |                                      |
| id                          | b574bf99-7877-4dd1-9f55-c6c984b39023 |
| insert_headers              | None                                 |
| l7policies                  |                                      |
| loadbalancers               | 1d76140c-de29-462a-bbb8-3fa98a5a3425 |
| name                        | vport1                               |
| operating_status            | OFFLINE                              |
| project_id                  | 3d21c71c4e4040599933bda62a76ae2f     |
| protocol                    | HTTP                                 |
| protocol_port               | 80                                   |
| provisioning_status         | PENDING_CREATE                       |
| sni_container_refs          | []                                   |
| timeout_client_data         | 50000                                |
| timeout_member_connect      | 5000                                 |
| timeout_member_data         | 50000                                |
| timeout_tcp_inspect         | 0                                    |
| updated_at                  | None                                 |
| client_ca_tls_container_ref | None                                 |
| client_authentication       | NONE                                 |
| client_crl_container_ref    | None                                 |
+-----------------------------+--------------------------------------+
```

ACOS log:
```
slb virtual-server 1d76140c-de29-462a-bbb8-3fa98a5a3425 192.168.91.56
  port 80 http
    name b574bf99-7877-4dd1-9f55-c6c984b39023
    conn-limit 20000
    extended-stats
    source-nat auto
!

```
---------------------

 - partition support
```
stack@ytsai-openstack:~/source/a10-octavia/device-flavor$ openstack loadbalancer create --flavor f_dev1 --vip-subnet-id tp91 --vip-address 192.168.91.56 --name vip1
+---------------------+--------------------------------------+
| Field               | Value                                |
+---------------------+--------------------------------------+
| admin_state_up      | True                                 |
| created_at          | 2021-05-12T12:41:55                  |
| description         |                                      |
| flavor_id           | cea57cae-abb4-4a57-824d-d7b2980395f6 |
| id                  | 2b2c83c2-f3a4-4871-9320-b22995f606f0 |
| listeners           |                                      |
| name                | vip1                                 |
| operating_status    | OFFLINE                              |
| pools               |                                      |
| project_id          | 3d21c71c4e4040599933bda62a76ae2f     |
| provider            | a10                                  |
| provisioning_status | PENDING_CREATE                       |
| updated_at          | None                                 |
| vip_address         | 192.168.91.56                        |
| vip_network_id      | 5e1b43e5-9f65-40af-a28f-88e63dc6d667 |
| vip_port_id         | 2b6d414d-33b6-4e26-9c7f-6c4078131a0b |
| vip_qos_policy_id   | None                                 |
| vip_subnet_id       | f25ce642-f953-4058-8c46-98fdd72fb129 |
+---------------------+--------------------------------------+
stack@ytsai-openstack:~/source/a10-octavia/device-flavor$ openstack loadbalancer listener create --protocol HTTP --protocol-port 80 --name vport1 vip1
+-----------------------------+--------------------------------------+
| Field                       | Value                                |
+-----------------------------+--------------------------------------+
| admin_state_up              | True                                 |
| connection_limit            | -1                                   |
| created_at                  | 2021-05-12T12:42:20                  |
| default_pool_id             | None                                 |
| default_tls_container_ref   | None                                 |
| description                 |                                      |
| id                          | ed519c5d-c437-40dc-81d4-4da8bc16a845 |
| insert_headers              | None                                 |
| l7policies                  |                                      |
| loadbalancers               | 2b2c83c2-f3a4-4871-9320-b22995f606f0 |
| name                        | vport1                               |
| operating_status            | OFFLINE                              |
| project_id                  | 3d21c71c4e4040599933bda62a76ae2f     |
| protocol                    | HTTP                                 |
| protocol_port               | 80                                   |
| provisioning_status         | PENDING_CREATE                       |
| sni_container_refs          | []                                   |
| timeout_client_data         | 50000                                |
| timeout_member_connect      | 5000                                 |
| timeout_member_data         | 50000                                |
| timeout_tcp_inspect         | 0                                    |
| updated_at                  | None                                 |
| client_ca_tls_container_ref | None                                 |
| client_authentication       | NONE                                 |
| client_crl_container_ref    | None                                 |
+-----------------------------+--------------------------------------+
```

```
AX1030[p7](config)#show running-config
!Current configuration: 288 bytes
!Configuration last updated at 20:57:05 CST Wed May 12 2021
!Configuration last saved at 00:27:14 CST Thu Apr 1 2021
!
active-partition p7
!
...
slb virtual-server 2b2c83c2-f3a4-4871-9320-b22995f606f0 192.168.91.56
  port 80 http
    name ed519c5d-c437-40dc-81d4-4da8bc16a845
    conn-limit 20000
    extended-stats
    source-nat auto
!
```

---------------------

 - HMT

```
stack@ytsai-openstack:~/source/a10-octavia/device-flavor$ openstack project list
+----------------------------------+--------------------+
| ID                               | Name               |
+----------------------------------+--------------------+
| 2824d8e196c640af9b3863c3d9b7b0d8 | alt_demo           |
| 2aae8b6e8ddd4e2991b6c342824056b0 | demo               |
| 3d21c71c4e4040599933bda62a76ae2f | admin              |
| 9955a18b75fa4c30bb865d72b5a86a6a | child1             |
| a9dbace62afb483898e7c4077d8c75d6 | invisible_to_admin |
| b612261c15374f1a896f4c3830042f48 | child2             |
| c377bd3ba07e43898018ef4f27b0ce0c | project_b          |
| c45ca1e789844ca0a0ea14da21d9a97e | service            |
| d97d24bfc7584be28b56acbdb793d89e | project_a          |
+----------------------------------+--------------------+

```

```
AX1030(config)#show partition
Total Number of active partitions: 2
Partition Name   Id     L3V/SP     Parent L3V           App Type   Admin Count
------------------------------------------------------------------------------
p7               7       L3V       -                    -            0
9955a18b75fa4c   8       L3V       -                    -            0
AX1030(config)#act
AX1030(config)#active-partition 9955a18b75fa4c
Current active partition: 9955a18b75fa4c
AX1030[9955a18b75fa4c](config)#show run
!Current configuration: 0 bytes
!Configuration last updated at 22:03:36 CST Wed May 12 2021
!Configuration last saved at 08:00:00 CST Thu Jan 1 1970
!
active-partition 9955a18b75fa4c
!
!
!
slb virtual-server 04ae6408-e5cd-4578-a736-7a49635d17d3 192.168.91.56
  port 80 http
    name 40102dc9-61ad-46b2-b330-b24ad8a790f6
    conn-limit 20000
    extended-stats
    source-nat auto
!
end
!Current config commit point for partition 8 is 0 & config mode is classical-mode

```

---------------------

- use parent partition

```
AX1030(config)#show partition
Total Number of active partitions: 3
Partition Name   Id     L3V/SP     Parent L3V           App Type   Admin Count
------------------------------------------------------------------------------
p7               7       L3V       -                    -            0
9955a18b75fa4c   8       L3V       -                    -            0
3d21c71c4e4040   9       L3V       -                    -            0
AX1030(config)#ac
AX1030(config)#act
AX1030(config)#active-partition 3d21c71c4e4040
Current active partition: 3d21c71c4e4040
AX1030[3d21c71c4e4040](config)#show run
AX1030[3d21c71c4e4040](config)#show running-config
!Current configuration: 0 bytes
!Configuration last updated at 22:10:43 CST Wed May 12 2021
!Configuration last saved at 08:00:00 CST Thu Jan 1 1970
!
active-partition 3d21c71c4e4040
!
!
!
slb virtual-server ae443a5d-519b-4130-8bb0-04064c5f327c 192.168.91.56
  port 80 http
    name f3103538-b9ea-4889-afe3-d4160d1befda
    conn-limit 20000
    extended-stats
    source-nat auto
!
end
!Current config commit point for partition 9 is 0 & config mode is classical-mode
```

### CODE REVIEW

File: a10_octavia/common/utils.py [hunk 1]
```diff
@@ -91,23 +91,36 @@ def convert_to_hardware_thunder_conf(hardware_list):
...
+            hardware_dict[project_id] = vthunder_conf
+
+        # Add dict entries with dev_name as key
+        device_name = hardware_device['device_name']
+        if hardware_dict.get(a10constants.DEVICE_KEY_PREFIX + device_name):
+            raise cfg.ConfigFileValueError('Supplied duplicate device_name {} '
+                                           ' in [hardware_thunder] section'.format(device_name))
         vthunder_conf = data_models.HardwareThunder(**hardware_device)
-        hardware_dict[project_id] = vthunder_conf
+        vthunder_conf.device_name_as_key = True
```
Comment (hthompson-a10) [inline a10_octavia/common/utils.py:121]: Maybe indentation is weird on github.  Looks like we are always setting `device_name_as_key` as true. Which means the check duplicate logic for HMT is never called.

Comment (hthompson-a10) [reply a10_octavia/common/utils.py:121]: To follow onto this, we don't know that they want to use the device name for creation until they actively use the flavor.

**Review by hthompson-a10 — CHANGES_REQUESTED** [review]

File: a10_octavia/common/utils.py (hunk 1 above)
Comment (hthompson-a10) [reply a10_octavia/common/utils.py:121]: Missed the iteration during review. nvm

File: a10_octavia/common/exceptions.py [hunk 2]
```diff
@@ -226,3 +226,16 @@ class NoFreeSetId(acos_errors.ACOSException):
     def __init__(self):
         msg = ('No free VRRP-A set_id available.')
         super(NoFreeSetId, self).__init__(msg=msg)
+
+
+class ProjectDeviceNotFound(acos_errors.ACOSException):
+    def __init__(self):
+        msg = ('No valid device found for this project, please check your configuration file.')
+        super(NoFreeSetId, self).__init__(msg=msg)
```
[the remark this answers is not in the document]
Comment (ytsai-a10) [reply a10_octavia/common/exceptions.py:234]: fixed

File: a10_octavia/controller/worker/tasks/vthunder_tasks.py [hunk 3]
```diff
@@ -1120,3 +1120,25 @@ class VthunderInstanceBusy(VThunderBaseTask):
...
+            if device_flavor is not None:
+                dev_key = a10constants.DEVICE_KEY_PREFIX + device_flavor
+                if dev_key in device_config_dict:
+                    vthunder_config = device_config_dict[dev_key]
+                    vthunder_config.project_id = loadbalancer.project_id
+                    if vthunder_config.hierarchical_multitenancy == "enable":
+                        vthunder_config.partition_name = loadbalancer.project_id[0:14]
+                    return vthunder_config, True
+                else:
+                    raise exceptions.FlavorDevieNotFound(device_flavor)
```
[the remark this answers is not in the document]
Comment (ytsai-a10) [reply a10_octavia/controller/worker/tasks/vthunder_tasks.py:1142]: fixed

File: a10_octavia/controller/worker/tasks/vthunder_tasks.py [hunk 4]
```diff
@@ -1120,3 +1120,25 @@ class VthunderInstanceBusy(VThunderBaseTask):
     def execute(self, compute_busy=False):
         if compute_busy:
             raise Exception('vThunder instance is busy now, try again later.')
+
+
+class GetVthunderConfByFlavor(VThunderBaseTask):
+
+    def execute(self, loadbalancer, vthunder_config, device_config_dict, flavor_data=None):
```
[the remark this answers is not in the document]
Comment (ytsai-a10) [reply a10_octavia/controller/worker/tasks/vthunder_tasks.py:1127]: added

**Review by hthompson-a10 — APPROVED** [review]

**Review by NehaKembalkarA10 — APPROVED** [review]

**Pull request merged**